File: kinotic-domain/src/main/java/org/kinotic/domain/api/model/WatchEvent.java
package org.kinotic.domain.api.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Date;

/**
 * One entry of the ledger: what happened to a watched record, from where, and why. The record itself
 * says what it is now; its entries say how it got there.
 *
 * @param timestamp    when the write landed, the data stream's time field
 * @param type         the kind of record that changed
 * @param id           the record's id
 * @param scope        the scope the record is addressed under, as its repository's {@code scopeOf} gives
 *                     it: its organization for a record an organization owns, null for a node
 * @param parent       what the record belongs to, null for a record the platform made on its own
 * @param kind         what happened
 * @param source       what caused it: the node that reported, the operator, the orchestrator's inference
 * @param serverNodeId the server node that wrote it
 * @param generation   the record's generation when written, null for a record that carries none
 * @param message      why, for an operator
 * @param value        what was written, as an object
 */
public record WatchEvent(@JsonProperty("@timestamp") Date timestamp,
                         WatchedType type,
                         String id,
                         String scope,
                         WatchedParent parent,
                         WatchEventKind kind,
                         String source,
                         String serverNodeId,
                         Long generation,
                         String message,
                         Object value) {
}

File: kinotic-js/workspace/packages/management-api/src/api/model/reconcile/WatchEvent.ts
import type { WatchEventKind } from './WatchEventKind'
import type { WatchedParent } from './WatchedParent'
import type { WatchedType } from './WatchedType'

/**
 * One entry of the ledger: what happened to a watched record, from where, and why. The record itself
 * says what it is now; its entries say how it got there.
 */
export interface WatchEvent {
    /** When the write landed, epoch milliseconds. */
    '@timestamp': number
    /** The kind of record that changed. */
    type: WatchedType
    /** The record's id. */
    id: string
    /**
     * The scope the record is addressed under, as its repository gives it: its organization for a
     * record an organization owns, null for a node.
     */
    scope: string | null
    /** What the record belongs to, null for a record the platform made on its own. */
    parent: WatchedParent | null
    /** What happened. */
    kind: WatchEventKind
    /** What caused it: the node that reported, the operator, the orchestrator's inference. */
    source: string
    /** The server node that wrote it. */
    serverNodeId: string
    /** The record's generation when written, null for a record that carries none. */
    generation: number | null
    /** Why, for an operator. */
    message: string
    /** What was written. */
    value: unknown
}
